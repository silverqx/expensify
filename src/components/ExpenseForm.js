import React, { Component } from 'react'

import moment from 'moment'
// TODO check this along with import 'react-dates/lib/css/_datepicker.css' in App.js silver
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import debounce from 'lodash.debounce'

import FormErrors from './FormErrors'

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props)

        const expense = props.expense || {}

        this.state = {
            expense: {
                description: expense.description || '',
                amount: expense.amount ? expense.amount + '' : '',
                createdAt: expense.createdAt ? moment(expense.createdAt) : moment(),
                note: expense.note || ''
            },
            datepickerFocused: false,
            errors: []
        }
    }

    componentWillUnmount() {
        this.validateFormDebounced.cancel()
    }

    onDescriptionChange = (e) => {
        const description = e.target.value

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                description
            }
        }))

        this.validateFormDebounced()
    }

    onAmountChange = (e) => {
        const amount = e.target.value

        // TODO take comma into account, eg try sk locale silver
        if (amount && !amount.match(/^\d+((\.|\,)\d{0,2})?$/))
            return

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                amount
            }
        }))

        this.validateFormDebounced()
    }

    onDateChange = (createdAt) => {
        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                createdAt
            }
        }))
    }

    onNoteChange = (e) => {
        const note = e.target.value

        this.setState((prevState) => ({
            expense: {
                ...prevState.expense,
                note
            }
        }))
    }

    validateFormDebounced = debounce(
        () => this.validateForm(), 600
    )

    validateForm = () => {
        const expense = this.state.expense
        let errors = []

        if (!expense.description)
            errors = errors.concat('Description is required.')
        if (!expense.amount)
            errors = errors.concat('Amount is required.')
        if (!expense.createdAt)
            errors = errors.concat('CreatedAt is required.')

        this.setState(() => ({ errors }))

        return !!!errors.length
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.validateForm())
            return

        const expense = this.state.expense

        this.props.onSubmit({
            ...expense,
            amount: parseFloat(expense.amount, 10) || 0,
            createdAt: expense.createdAt.valueOf()
        })
    }

    render() {
        return (
            <div>
                <FormErrors errors={this.state.errors} />
                <form onSubmit={this.onSubmit}>
                    <div>
                        <input
                            type="text"
                            autoFocus
                            maxLength={255}
                            placeholder="Description..."
                            value={this.state.expense.description}
                            onChange={this.onDescriptionChange}
                        />
                    </div>
                    <div>
                        <input
                            type="number"
                            placeholder="Amount..."
                            step={1.0}
                            value={this.state.expense.amount}
                            onChange={this.onAmountChange}
                            step={0.5}
                        />
                    </div>
                    <div>
                        <SingleDatePicker
                            id="created-at"
                            date={this.state.expense.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.datepickerFocused}
                            onFocusChange={({ focused }) => this.setState({ datepickerFocused: focused })}
                            isOutsideRange={() => false}
                            numberOfMonths={1}
                            placeholder="Created At..."
                            small={true}
                            daySize={32}
                        />
                    </div>
                    <div>
                        <textarea
                            cols="30"
                            rows="4"
                            placeholder="Note... (optional)"
                            value={this.state.expense.note}
                            onChange={this.onNoteChange}
                        ></textarea>
                    </div>
                    <div>
                        <input
                            type="submit"
                            value={this.props.expense ? 'Edit Expense' : 'Add Expense'}
                        />
                    </div>
                </form>
            </div>
        )
    }
}
