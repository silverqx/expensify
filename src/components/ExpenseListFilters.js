import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

// TODO clear input on esc, use hoc https://github.com/amytych/react-escape-outside/blob/master/index.js
export class ExpenseListFilters extends Component {
    state = {
        datepickerFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    getSortFilter = (value) => {
        const mapValueToFilter = {
            date: 'sortByDate',
            amount: 'sortByAmount'
        }
        const sortFilter = mapValueToFilter[value]

        return this.props[sortFilter]()
    }

    render() {
        return (
            <div>
                <input
                    autoFocus
                    type="text"
                    value={this.props.filters.text}
                    onChange={e => this.props.setTextFilter(e.target.value)}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={e => this.getSortFilter(e.target.value)}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    startDateId="created-at-start-date"
                    endDateId="created-at-end-date"
                    startDatePlaceholderText="Start Date..."
                    endDatePlaceholderText="End Date..."
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    small={true}
                    daySize={32}
                    showClearDates={true}
                    focusedInput={this.state.datepickerFocused}
                    onDatesChange={this.onDatesChange}
                    onFocusChange={datepickerFocused => this.setState({ datepickerFocused })}
                />
            </div>
        )
    }
}

const mapStateToProps = ({ filters }) => ({
    filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ExpenseListFilters)
