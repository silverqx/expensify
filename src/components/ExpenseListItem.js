import React from 'react'
import { connect } from 'react-redux'
import moment from 'moment'
import { Link } from 'react-router-dom'

import { removeExpense } from '../actions/expenses'

// TODO find out how to correctly pass history and expense silver
const ExpenseListItem = ({ dispatch, history, id, description, note, amount, createdAt}) => (
    <div>
        <Link to={`/edit/${id}`}>{description}</Link> - {amount} - {moment(createdAt).format('LL')}
        <button
            onClick={() => history.push(`/edit/${id}`)}
        >
            Edit
        </button>
        <button
            onClick={() => dispatch(removeExpense(id))}
        >
            Remove
        </button>
    </div>
)

export default connect()(ExpenseListItem)
