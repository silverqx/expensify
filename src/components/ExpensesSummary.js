import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import 'numeral/locales/sk'

import getExpensesTotal from '../selectors/expenses-total'
import getVisibleExpenses from '../selectors/expenses'

export const ExpensesSummary = ({ expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
    const formattedExpensesTotal = numeral(expensesTotal).format('$0,0.00')

    return (
        <div>
            <div>
                <span>Showing {expensesCount} {expenseWord} </span>
                <span>with total amount {formattedExpensesTotal}</span>
            </div>
        </div>
    )
}

const mapStateToProps = ({ expenses, filters }) => {
    const visibleExpenses = getVisibleExpenses(expenses, filters)

    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)
