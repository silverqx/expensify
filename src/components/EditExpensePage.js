import React, { Component } from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'

import { editExpense } from '../actions/expenses'

export class EditExpensePage extends Component {
    onSubmit = (updatedExpense) => {
        this.props.editExpense(
            this.props.match.params.id,
            updatedExpense
        )
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

// FIXME really bad solution, should be done in router? And how to test this? silver
const findExpense = (state, { history, match }) => {
    const expense = state.expenses.find(
        expense => expense.id === match.params.id
    )

    if (!expense)
        return history.goBack()

    return expense
}

const mapStateToProps = (state, props) => ({
    expense: findExpense(state, props)
})

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updatedExpense) => dispatch(
        editExpense(id, updatedExpense)
    )
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage)
