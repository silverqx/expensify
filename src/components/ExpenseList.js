import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'

import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = ({ history, expenses }) => (
    <div className="content-container">
        <div className="list">
            <div className="list__header">
                <div className="show-on-desktop">Expense</div>
                <div className="show-on-desktop">Amount</div>
                <div className="show-on-mobile">Expenses</div>
            </div>

            <div className="list__content">
                {
                    expenses.length === 0 ?
                        <div className="list__content--empty">No Expenses</div> :
                        expenses.map((expense) => (
                            <ExpenseListItem
                                key={expense.id}
                                {...expense}
                            />
                        ))
                }
            </div>
        </div>
    </div>
)

const mapStateToProps = ({ expenses, filters }) => ({
    expenses: getVisibleExpenses(expenses, filters)
})

export default connect(mapStateToProps)(ExpenseList)
