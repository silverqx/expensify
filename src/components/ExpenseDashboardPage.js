import React from 'react'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <h1>Expense Dashboard Page</h1>
        <h3>Expenses List</h3>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage
