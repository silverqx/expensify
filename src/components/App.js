import React, { Component } from 'react'
import { Router, Route, Switch } from "react-router-dom"
import ReactModal from 'react-modal'
import { createBrowserHistory } from 'history'

import ExpenseDashboardPage from './ExpenseDashboardPage'
import AddExpensePage from './AddExpensePage'
import EditExpensePage from './EditExpensePage'
import LoginPage from './LoginPage'
import NotFoundPage from './NotFoundPage'

import ModalOption from './ModalOption'
import PrivateRoute from '../routers/PrivateRoute'
import PublicRoute from '../routers/PublicRoute'

// import { clearInputValue } from '../utils/utils'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

// Make sure to bind modal to your appElement
// (http://reactcommunity.org/react-modal/accessibility/)
ReactModal.setAppElement('#app')

export const history = createBrowserHistory()

export default class ExpensifyApp extends Component {
    static defaultProps = {
        title: 'Expensify'
    }

    state = {
        error: null,
        // TODO implement subtitle silver
        subtitle: 'Xxx',
    }

    componentDidMount() {
    }

    componentDidUpdate(prevProps, prevState) {
    }

    /**
     * Update Error state.
     *
     * @param {string} error - Error message.
     *
     * @returns {?boolean} True every time.
     */
    handleError = (error) => {
        this.setState(() => ({ error }))

        return true
    }

    /**
     * Clear Error state.
     *
     * @returns {?boolean} False every time.
     */
    clearError = () => {
        this.setState(() => ({ error: null}))

        return false
    }

    render() {
        return (
            <Router history={history}>
                <Switch>
                    <PublicRoute path="/" exact component={LoginPage} />
                    <PrivateRoute path="/dashboard" component={ExpenseDashboardPage} />
                    <PrivateRoute path="/create" component={AddExpensePage} />
                    <PrivateRoute path="/edit/:id" component={EditExpensePage} />
                    <Route component={NotFoundPage} />
                </Switch>
                <ModalOption
                    selectedOption={this.state.selectedOption}
                    handleClearSelectedOption={this.handleClearSelectedOption}
                />
            </Router>
        )
    }
}
