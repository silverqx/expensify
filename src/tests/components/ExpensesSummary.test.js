import React from 'react'
import { shallow } from 'enzyme'

import { ExpensesSummary } from '../../components/ExpensesSummary'

describe('ExpensesSummary component', () => {
    test('should correctly render ExpensesSummary with 1 expense', () => {
        const wrapper = shallow(
            <ExpensesSummary
                expensesCount={1}
                expensesTotal={50.25}
            />)

        expect(wrapper).toMatchSnapshot()
    })

    test('should correctly render ExpensesSummary multiple expenses', () => {
        const wrapper = shallow(
            <ExpensesSummary
                expensesCount={5}
                expensesTotal={244.32}
            />)

        expect(wrapper).toMatchSnapshot()
    })
})
