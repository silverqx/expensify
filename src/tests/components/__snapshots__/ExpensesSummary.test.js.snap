// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpensesSummary component should correctly render ExpensesSummary multiple expenses 1`] = `
<div>
  <div>
    <span>
      Showing 
      5
       
      expenses
       
    </span>
    <span>
      with total amount 
      $244.32
    </span>
  </div>
</div>
`;

exports[`ExpensesSummary component should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <div>
    <span>
      Showing 
      1
       
      expense
       
    </span>
    <span>
      with total amount 
      $50.25
    </span>
  </div>
</div>
`;
